src/prefect/utilities/executors.py

```diff
@@ -176,81 +216,128 @@ def multiprocessing_timeout(
+    kwargs: Mapping = None,
+    logger: Logger = None,
 ) -> Any:
     """
-    Helper function for implementing timeouts on function executions.
+    Helper function for implementing timeouts on task executions.
 
     The exact implementation varies depending on whether this function is being
     run in the main thread or a non-daemonic subprocess.  If this is run from a
-    daemonic subprocess or on Windows, the task is run in a
-    `ThreadPoolExecutor` and only a soft timeout is enforced, meaning a
-    `TimeoutError` is raised at the appropriate time but the task continues
-    running in the background.
+    daemonic subprocess or on Windows, the task is run in a `ThreadPoolExecutor`
+    and only a soft timeout is enforced, meaning a `TimeoutError` is raised at the
+    appropriate time but the task continues running in the background.
+
+    The task is passed instead of a function so we can give better logs and messages.
+    If you need to run generic functions with timeout handlers,
+    `run_with_thread_timeout` or `run_with_multiprocess_timeout` can be called directly
 
     Args:
-        - fn (callable): the function to execute
-        - *args (Any): arguments to pass to the function
-        - timeout (int): the length of time to allow for execution before raising a
-            `TimeoutError`, represented as an integer in seconds
-        - **kwargs (Any): keyword arguments to pass to the function
+        - task (Task): the task to execute
+            `task.timeout` specifies the number of seconds to allow `task.run` to run
+            for before terminating
+        - args (Sequence): arguments ``to pass to the function
+        - kwargs (Mapping): keyword arguments to pass to the function
+        - logger (Logger): an optional logger to use. If not passed, a logger for the
+            `prefect.run_task_with_timeout_handler` namespace will be created.
 
     Returns:
         - the result of `f(*args, **kwargs)`
 
     Raises:
         - TimeoutError: if function execution exceeds the allowed timeout
     """
+    logger = logger or get_logger("run_task_with_timeout_handler")
+    name = prefect.context.get("task_full_name", task.name)
+    kwargs = kwargs or {}
+
     # if no timeout, just run the function
-    if timeout is None:
-        return fn(*args, **kwargs)
+    if task.timeout is None:
+        return task.run(*args, **kwargs)  # type: ignore
 
     # if we are running the main thread, use a signal to stop execution at the
     # appropriate time; else if we are running in a non-daemonic process, spawn
     # a subprocess to kill at the appropriate time
     if not sys.platform.startswith("win"):
+
         if threading.current_thread() is threading.main_thread():
-            return main_thread_timeout(fn, *args, timeout=timeout, **kwargs)
+            # This case is typically when using the LocalDaskExecutor with a
+            # process-based scheduler because then each worker is in the main thread
```

Comment:
My intention here was to cover that the thread based timeout is being used when your executor is process based which is non-intuitive and lead me to debug the wrong thing for a while. I think at least saying something like...

- If the task is running in the main thread, use a `signal.alarm` based timeout. This is typically seen when the executor is not using multiple threads.
- If the task is running in a non-daemonic subprocess, use a subprocess based timeout. This is typically seen with the multithreaded distributed executor.

Reply:
I agree a list of executors isn't ideal

Reply:
I'm fine with describing the logic here, I'd just remove any reference to a specific executor by name

Reply:
I removed specific references but I think I'm still kind of identifying executors that would cause it so this may not accomplish what you wanted still, lmk.

Reply:
LGTM, thanks